File: src/GameState.js
import { Node } from './Node.js';

/**
 * Clase que maneja el estado del juego y las reglas
 */
export class GameState {
    constructor(level, robotX, robotY) {
        this.level = level;
        this.rows = level.length;
        this.cols = level[0].length;
        this.robotX = robotX;
        this.robotY = robotY;
        
        // Encontrar posiciones de las luces
        this.lightPositions = [];
        for (let i = 0; i < this.rows; i++) {
            for (let j = 0; j < this.cols; j++) {
                if (level[i][j] === 2) {
                    this.lightPositions.push([i, j]);
                }
            }
        }
        
        // Estado inicial: todas las luces apagadas
        this.initialLights = new Array(this.lightPositions.length).fill(0);
    }

    /**
     * Verifica si una posición está dentro del tablero
     */
    isValidPosition(x, y) {
        return x >= 0 && x < this.rows && y >= 0 && y < this.cols;
    }

    /**
     * Verifica si el robot puede moverse a una posición
     */
    canMoveTo(x, y) {
        return this.isValidPosition(x, y) && this.level[x][y] !== 1;
    }

    /**
     * Verifica si el robot puede encender una luz en su posición actual
     */
    canTurnOnLight(x, y, lights) {
        // Buscar si hay una luz en esta posición
        const lightIndex = this.lightPositions.findIndex(([lx, ly]) => lx === x && ly === y);
        
        // Puede encender si hay una luz y está apagada
        return lightIndex !== -1 && lights[lightIndex] === 0;
    }

    /**
     * Genera todos los sucesores posibles de un nodo
     */
    getSuccessors(node) {
        const successors = [];
        const directions = [
            [-1, 0, 'ARRIBA'],
            [1, 0, 'ABAJO'],
            [0, -1, 'IZQUIERDA'],
            [0, 1, 'DERECHA']
        ];

        // Intentar movimientos en las 4 direcciones
        for (const [dx, dy, action] of directions) {
            const newX = node.x + dx;
            const newY = node.y + dy;

            if (this.canMoveTo(newX, newY)) {
                const successor = new Node(
                    newX, 
                    newY, 
                    [...node.lights], 
                    node, 
                    action, 
                    node.cost + 1
                );
                successors.push(successor);
            }
        }

        // Intentar encender luz en la posición actual
        if (this.canTurnOnLight(node.x, node.y, node.lights)) {
            const lightIndex = this.lightPositions.findIndex(([lx, ly]) => lx === node.x && ly === node.y);
            const newLights = [...node.lights];
            newLights[lightIndex] = 1;

            const successor = new Node(
                node.x, 
                node.y, 
                newLights, 
                node, 
                'ENCENDER', 
                node.cost + 1
            );
            successors.push(successor);
        }

        return successors;
    }

    /**
     * Verifica si un nodo representa el estado meta (todas las luces encendidas)
     */
    isGoal(node) {
        return node.lights.every(light => light === 1);
    }

    /**
     * Calcula la heurística para A*
     * Heurística = número de luces apagadas + distancia a la luz más cercana
     */
    heuristic(node) {
        const lightsOff = node.lights.filter(light => light === 0).length;
        
        if (lightsOff === 0) {
            return 0;
        }

        // Encontrar distancia a la luz apagada más cercana
        let minDistance = Infinity;
        
        for (let i = 0; i < this.lightPositions.length; i++) {
            if (node.lights[i] === 0) {
                const [lx, ly] = this.lightPositions[i];
                const distance = Math.abs(node.x - lx) + Math.abs(node.y - ly);
                minDistance = Math.min(minDistance, distance);
            }
        }

        return lightsOff + minDistance;
    }

    /**
     * Crea el nodo inicial
     */
    getInitialNode() {
        return new Node(this.robotX, this.robotY, [...this.initialLights]);
    }
}

File: game_state.py
"""
Clase que maneja el estado del juego y las reglas
"""
from node import Node

class GameState:
    def __init__(self, level, robot_x, robot_y):
        self.level = level
        self.rows = len(level)
        self.cols = len(level[0])
        self.robot_x = robot_x
        self.robot_y = robot_y
        
        # Encontrar posiciones de las luces
        self.light_positions = []
        for i in range(self.rows):
            for j in range(self.cols):
                if level[i][j] == 2:
                    self.light_positions.append((i, j))
        
        # Estado inicial: todas las luces apagadas
        self.initial_lights = tuple([0] * len(self.light_positions))

    def is_valid_position(self, x, y):
        """Verifica si una posición está dentro del tablero"""
        return 0 <= x < self.rows and 0 <= y < self.cols

    def can_move_to(self, x, y):
        """Verifica si el robot puede moverse a una posición"""
        return self.is_valid_position(x, y) and self.level[x][y] != 1

    def can_turn_on_light(self, x, y, lights):
        """Verifica si el robot puede encender una luz en su posición actual"""
        # Buscar si hay una luz en esta posición
        light_index = None
        for i, (lx, ly) in enumerate(self.light_positions):
            if lx == x and ly == y:
                light_index = i
                break
        
        # Puede encender si hay una luz y está apagada
        return light_index is not None and lights[light_index] == 0

    def get_successors(self, node):
        """Genera todos los sucesores posibles de un nodo"""
        successors = []
        directions = [
            (-1, 0, 'ARRIBA'),
            (1, 0, 'ABAJO'),
            (0, -1, 'IZQUIERDA'),
            (0, 1, 'DERECHA')
        ]

        # Intentar movimientos en las 4 direcciones
        for dx, dy, action in directions:
            new_x = node.x + dx
            new_y = node.y + dy

            if self.can_move_to(new_x, new_y):
                successor = Node(
                    new_x, 
                    new_y, 
                    list(node.lights), 
                    node, 
                    action, 
                    node.cost + 1
                )
                successors.append(successor)

        # Intentar encender luz en la posición actual
        if self.can_turn_on_light(node.x, node.y, node.lights):
            light_index = None
            for i, (lx, ly) in enumerate(self.light_positions):
                if lx == node.x and ly == node.y:
                    light_index = i
                    break
            
            new_lights = list(node.lights)
            new_lights[light_index] = 1

            successor = Node(
                node.x, 
                node.y, 
                new_lights, 
                node, 
                'ENCENDER', 
                node.cost + 1
            )
            successors.append(successor)

        return successors

    def is_goal(self, node):
        """Verifica si un nodo representa el estado meta (todas las luces encendidas)"""
        return all(light == 1 for light in node.lights)

    def heuristic(self, node):
        """
        Calcula la heurística para A*
        Heurística = número de luces apagadas + distancia a la luz más cercana
        """
        lights_off = sum(1 for light in node.lights if light == 0)
        
        if lights_off == 0:
            return 0

        # Encontrar distancia a la luz apagada más cercana
        min_distance = float('inf')
        
        for i, (lx, ly) in enumerate(self.light_positions):
            if node.lights[i] == 0:
                distance = abs(node.x - lx) + abs(node.y - ly)
                min_distance = min(min_distance, distance)

        return lights_off + min_distance

    def get_initial_node(self):
        """Crea el nodo inicial"""
        return Node(self.robot_x, self.robot_y, list(self.initial_lights))